gwctl/pkg/resourcediscovery/discoverer.go

```diff
@@ -421,6 +449,103 @@ func (d Discoverer) discoverHTTPRoutesFromBackends(ctx context.Context, resource
+	// This will be a three step process:
+	//  1. Add ALL Backends to the resourceModel that are referenced by the
+	//     HTTPRoutes.
+	//  2. Discover ReferenceGrants for those Backends.
+	//  3. Remove Backends from the resourceModel which are in a different namespace
+	//     from the HTTPRoute and are not exposed through a ReferenceGrant.
+
+	// Step 1
+	for _, httpRouteNode := range resourceModel.HTTPRoutes {
+		for _, backendRef := range relations.FindBackendRefsForHTTPRoute(*httpRouteNode.HTTPRoute) {
+			// Check if the Backend already exists in the resourceModel
+			backendID := BackendID(backendRef.Group, backendRef.Kind, backendRef.Namespace, backendRef.Name)
+			if _, ok := resourceModel.Backends[backendID]; ok {
+				// Backend already exists in the resourceModel, skip re-fetching.
+				continue
+			}
+
+			// Backend doesn't already exist so fetch and add it to the resourceModel.
+			backends, err := d.fetchBackends(ctx, Filter{Namespace: backendRef.Namespace, Name: backendRef.Name})
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					err := ReferenceToNonExistentResourceError{ReferenceFromTo: ReferenceFromTo{
+						ReferringObject: common.ObjRef{Kind: "HTTPRoute", Name: httpRouteNode.HTTPRoute.GetName(), Namespace: httpRouteNode.HTTPRoute.GetNamespace()},
+						ReferredObject:  backendRef,
+					}}
+					httpRouteNode.Errors = append(httpRouteNode.Errors, err)
+					klog.V(1).Info(err)
+				} else {
+					klog.V(1).ErrorS(err, "Error while fetching Backend for HTTPRoute",
+						"backend", fmt.Sprintf("%v/%v/%v", backendRef.Kind, backendRef.Namespace, backendRef.Name),
+						"httproute", httpRouteNode.HTTPRoute.GetNamespace()+"/"+httpRouteNode.HTTPRoute.GetName(),
+					)
+				}
+				continue
+			}
+			resourceModel.addBackends(backends[0])
+		}
+	}
+
+	// Step 2
+	d.discoverReferenceGrantsFromBackends(ctx, resourceModel)
+
+	// Step 3
+	for _, httpRouteNode := range resourceModel.HTTPRoutes {
+		for _, backendRef := range relations.FindBackendRefsForHTTPRoute(*httpRouteNode.HTTPRoute) {
+			// Check if the Backend exists in the resourceModel
+			backendID := BackendID(backendRef.Group, backendRef.Kind, backendRef.Namespace, backendRef.Name)
+			backendNode, ok := resourceModel.Backends[backendID]
+			if !ok {
+				// Backend does not exist in the resourceModel (maybe because of some
+				// error), so skip processing this backend.
+				continue
+			}
+
+			// Ensure that if this is a cross namespace reference, then it is accepted
+			// through some ReferenceGrant.
+			if httpRouteNode.HTTPRoute.GetNamespace() != backendRef.Namespace {
+				httpRouteRef := common.ObjRef{
+					Group:     httpRouteNode.HTTPRoute.GroupVersionKind().Group,
+					Kind:      httpRouteNode.HTTPRoute.GroupVersionKind().Kind,
+					Name:      httpRouteNode.HTTPRoute.GetName(),
+					Namespace: httpRouteNode.HTTPRoute.GetNamespace(),
+				}
+				var referenceAccepted bool
+				for _, referenceGrantNode := range backendNode.ReferenceGrants {
```

Comment:
It would probably be useful to have some basic scalability tests - let's say 100 of each kind of resources with different levels of connections between to help keep some basic understanding of the impact each additional change like this adds. Not a blocker for this PR, but could be useful as a follow up issue for someone to tackle.